gdal: update proj dependency

danielnachun opened the pull request:
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

carlocab commented:
Dependencies `libgeotiff` and `libspatialite` need to also be migrated. Along with everything that depends on `gdal`.

danielnachun commented:
We're in a bit of bind here.  I was able to migrate all dependents of `proj@7` to `proj` except 3:
- `liblwgeom` - deprecated, appears to be superseded by `librttopo`
- `osm2pgsql` - this project appears to be active (the last release was 16 days ago).  We should open an upstream issue to inquire about PROJ 8 support
- `spatialite-gui` - this is where our remaining dependency conflict comes from, and it's not pretty.  The last stable version doesn't support PROJ 8, but the 2.1 beta does.  I tried to build the 2.1 beta just to see if would work, and it fails because it needs `librasterlite2`, which we don't package.  Our `librasterlite` formula is for an older library which is no longer developed because it has been superseded by `librasterlite2`.  Potentially we could just update the `librasterlite` formula because it has no dependents.  But that doesn't help with version 2.1 of `spatialite-gui` still being in beta.

Any thoughts on the best solution here?

danielnachun commented:
I've looked into how Linux distros have handled `librasterlite`/`librastlerlite2` and it appears that both Debian and Fedora removed `librasterlite` and replaced it with `librasterlite2` some time ago.  So it seems that we are actually behind the times and should probably do the same.  It seems like a formula rename would make sense to match what other distros do.

I was looking at the project timeline for `spatialite-gui` and there seems to be quite recent development activity: https://www.gaia-gis.it/fossil/spatialite_gui/timeline.  However, the last stable version (1.7.1, which we are using) was released in **2013**.  The version `2.0.0-devel` was released in 2015, and since then there's been a `2.1.0-beta0` and `2.1.0-beta1`.  I think we should clarify with upstream why this project isn't following making stable releases before going to the next beta.

carlocab commented:
It could be that they just have a stricter definition of "stable" than the one we usually use. It might mean "bug fixes only, no changes at all to the API", which is not uncommon.

carlocab commented:
Debian and Fedora ship spatialite-gui 2.1.0-beta. I think we can do the same.

GeraldCNelson commented:
For what it is worth, here's the librasterlite website - https://www.gaia-gis.it/fossil/librasterlite/index. Two things to note. 1. is that it is no longer maintained and advises using librasterlite. 2. librasterlite and spatialite are done by the same author. spatialite guide website is here - https://www.gaia-gis.it/fossil/libspatialite/index. spatiallite proper is here - https://www.gaia-gis.it/fossil/libspatialite/index. Its current version is 5.01 released in 2021.

danielnachun commented:
> Debian and Fedora ship spatialite-gui 2.1.0-beta. I think we can do the same.

I was leaning towards doing that.  I'm going to formula rename of `librasterlite` to `librasterlite2`, update it to the latest version, and update `spatialite-gui` to `2.1.0-beta`.  

I'm also going to disable `proj` support in `osm2pgsql` because they recently added an option to do this https://github.com/openstreetmap/osm2pgsql/pull/1331/files, and apparently `libosmium`, which is upstream to `osm2pgsql`, considers PROJ support for `libosmium` deprecated: https://github.com/osmcode/libosmium/issues/295.  This will leave only `liblwgeom` as a `proj@7` dependent, and as mentioned above, `liblwgeom` is itself deprecated.

danielnachun commented:
I was able to successfully build both `librasterlite2` and `spatialiate-gui` locally for macOS and Linux.  However, while doing this testing I discovered a number of other issues that need to be fixed before I think we can move further with this:
- It turns out `libgaiagraphics` actually doesn't work with `proj`.  It found `proj@7` through opportunistic linking on macOS, but failed in the Linux VM because `proj@7` wasn't installed.  However, the upstream page for `libgaiagraphics` explicitly states it is deprecated and its replacement is... `librasterlite2`.  So instead of migrating `libgaiagraphics` to from `proj@7`, I will deprecate it.  Its only dependent was `spatiliate-gui`, which now uses `librasterlite2` instead, so we don't have to worry about the deprecation causing issues for any other dependents.
- `spatialiate-tools` was unbottled on Linux because it has an unbottled dependecy `readosm`, which had a broken test on Linux, and because it appends `-liconv` unconditionally to `LDFLAGS`, when this should be macOS-only.  I've opened https://github.com/Homebrew/homebrew-core/pull/94995 to fix `readosm` (which of course had some other issues too, though they were fixable), and I have a PR ready to fix `spatialite-tools` once `readosm` is bottled on Linux.
- `mapserver` was unbottled on Linux because it needs to be built with GCC 11.  I have opened https://github.com/Homebrew/homebrew-core/pull/94997 to fix this.
- `postgis` was unbottled on Linux because it also needs to be built with GCC 11, and there may also be an issue because has a weird install process after compilation that installs to staging directory instead of the keg, and then manually copies over some files.  I'm working on a fix for this now that is hopefully less messy.

Given how deeply this PR has dragged us into dependency hell, I'm inclined now to try to get `librasterlite2` merged and `spatialite-gui` updated while still having them build against `proj@7`, and then rebase this PR once all these other issues are fixed so that it can just be focused on the dependency migration.

danielnachun commented:
As suggested in https://github.com/Homebrew/homebrew-core/pull/95481, we will update `spatialite-gui` in this PR rather than doing it separately.   A few more updates:
- It turns out that `osm2pgsql` does support `proj`, but defaults to using `proj@7` if it finds it.  That caused it to fail to build due to opportunistic linking on my local installation.  In CI this won't happen, but `osm2pgsql` has added a `USE_PROJ_LIB` CMake flag that I have set to always use the newer PROJ API even if the old one is detected.
- We may have to disable `libgaiagraphics` here or in a separate PR, because it has what I believe is an unresolvable dependency conflict.  It has been deprecated upstream because it has been replaced by `librasterlite2` and will never be updated to support `proj`.  However, it depends on `libgeotiff`, which does support `proj` and is being updated in this PR.  We could add it the list of allowable dependency conflicts, but it could be broken and that seems as bad if not worse than just disabling it.
- `librasterlite` is not part of the dependency tree here (although it is a dependent of `libgeotiff` and `libspatialite`, but it should still be deprecated in a separate PR, because has also been replaced by `librasterlite2` and will not be updated if it has future dependency issues.

carlocab commented on Formula/osm2pgsql.rb:
```diff
@@ -31,13 +31,8 @@ def install
-
     mkdir "build" do
-      system "cmake", "-DWITH_LUAJIT=ON", "..", *std_cmake_args
+      system "cmake", "-DWITH_LUAJIT=ON", "-DUSE_PROJ_LIB=6", "..", *std_cmake_args
```
Where does the `6` come from?

danielnachun replied:
As described here: https://github.com/openstreetmap/osm2pgsql/commit/552e4bc0b874989878ea529193d106eeaed2e7e5, the old API for PROJ was actually implemented in PROJ 4, and the new API debuted in PROJ 6.  Both APIs were supported in PROJ 6 and 7, but in PROJ 8, the old PROJ API was removed.  So the number here is referring the version in which the API was first implemented, not the version currently being used.

cho-m commented:
May need to check compatibility with PROJ 9, which will be merged with #96103

danielnachun commented:
I'm going to disable `libgaiagraphics` because I don't believe there's a way us to ship a working bottle anymore, and I'll rebase this against PROJ 9.  I expect it should work, as the underlying API does not appear to have fundamentally changed since PROJ 6.

danielnachun commented:
It looks like we also have to disabled `librasterlite` since it links to `proj@7` but it depends on `libgeotiff` and `libspatialite` which are being upgrade to `proj`.  I don't believe its possible for `librasterlite` and `libgaiagraphics` to work anymore, but this shouldn't really matter to us because both are deprecated upstream and have been replaced by `librasterlite2`.

cho-m commented:
Looks like CI  wasn't able to clean up leftover installed formulae after hitting error in a previous run. The `libspatialite` brew doctor issue is showing up in other PRs, so may need someone to manually fix CI runner.
```
==> brew doctor
==> FAILED
Full doctor output
  Please note that these warnings are just used to help the Homebrew maintainers
  with debugging if you file an issue. If everything you use Homebrew for is
  working fine: please don't worry or file an issue; just ignore this. Thanks!
  
  Warning: You have unlinked kegs in your Cellar.
  Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
  those kegs to fail to run properly once built. Run `brew link` on these:
    libspatialite
```

danielnachun commented:
@carlocab I've updated this to deprecate `libgaiagraphics` and `librasterlite` instead of disabling them, and added them to the versioned dependency conflict allow list.  Unfortunately neither formula has a test and I'm not sure it's worth the work to add one for formulae we're about to deprecate.

cho-m commented:
`gdal` is missing a revision bump.

danielnachun commented:
> `gdal` is missing a revision bump.

Good catch, I fixed this now.

cho-m commented:
`librasterlite` build failure:
```
configure: error: 'libproj' is required but it doesn't seem to be installed on this system.
```

danielnachun commented:
It looks like the only thing needed for `librasterlite` to build was an explicit dependency on `proj@7`.  I think we're good to go!

cho-m commented on Formula/spatialite-gui.rb:
```diff
@@ -1,10 +1,9 @@
   license "GPL-3.0-or-later"
-  revision 9
 
   livecheck do
     url "https://www.gaia-gis.it/gaia-sins/spatialite-gui-sources/"
```
In separate PR, we may want to relax livecheck regex to allow for `-beta` versions to avoid warnings until 2.x is out of beta.

cho-m commented on Formula/spatialite-gui.rb:
```diff
@@ -24,43 +23,39 @@ class SpatialiteGui < Formula
-    sha256 "37f71f3cb2b0b9649eb85a51296187b0adf2972c5a1d3ee0daf3082e2c35025e"
-  end
+  uses_from_macos "curl"
+  uses_from_macos "libxml2"
```
Not too important, but Homebrew's `libxml2` is getting linked rather than using macOS version. Probably pulled in via `libspatialite`.

Shouldn't be an issue as indirect dependency avoids broken linkage. In future, may want to check if `libspatialite` truly needs a newer copy of `libxml2` than what is provided by system.

cho-m reviewed (approved):
I don't see any issues blocking merge.

Can wait for @carlocab to also take a look.

carlocab commented on Formula/spatialite-gui.rb:
```diff
@@ -24,43 +23,39 @@ class SpatialiteGui < Formula
   depends_on "sqlite"
+  depends_on "virtualpg"
+  depends_on "webp"
   depends_on "wxwidgets@3.0"
```
Did we check if this supports the latest `wxwidgets`?

danielnachun replied:
I forgot to mention this, but I tried earlier and it did not work.  In fact the AUR package for `spatialite-gui` is broken because they are trying to use the newest `wxwidgets` and I saw the same error when I tried to upgrade.
